In [1]:
#Question 1)
 
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [3]:
#Question 2.1)
 
df.shape #prints out the number of rows and columns 

(391, 11)

## Question 2.2)

General definitions of what observations and variables are:

**Observations:** Observations are the individual datapoints recorded. This would refer to the horizontal set (the row). 

**Variables:** Variables are the categories of recorded data. This would refer the the vertical set (the columns). 

In [5]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [6]:
#Question 3)

df['species'].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [7]:
#Question 4)

import pandas as pd

second_url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic = pd.read_csv(second_url)
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [8]:
titanic.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [8]:
titanic.shape

(891, 15)

## Question 4)
**Why are df.shape and df.describe not the same?**

**a)** The describe method only provides calculations for numerical variables (6 in total) while shape provides the total number of columns, regardless of the type of data (17). 

**b)** There are different numbers provided from both the describe method and shape attribute due to null values within the variable (column) "age". While testing the sum of null values within "age", we find 177 missing values. 

The shape attribute provides the total number of rows while describe counts the number of values present, thus providing a different number.

## Question 5)
**Attributes:** Attributes are not calling a function, and refer to information already within the object itself. Thus, they do not need to call a function and run code and do not need parenthesis. 

**Methods:** Methods call functions and run code based off of it. Thus, they need parenthesis. 

## Question 6)
**Definitions of what the method df.describe provides:**
1. count: tallies the number of values within the column of data
2. mean: calculates the average of all the values within the column
3. std: calulcates the standard deviation. Standard deviation represents how much the values are different from the mean. 
4. min: caluclates the smallest number
5. 25% is the first quartile, 50% is the second, 75% is the third and max is the fourth. 

### Chatbot log upto this point: 
Here’s a summary of our session so far:

1. **Dataset Overview**:
   - You started by asking about the dataset on Animal Crossing characters (villagers) that you downloaded. You wanted to know about the columns and the size of the dataset.

2. **Dataset Exploration in Jupyter**:
   - I provided Python code examples using `pandas` to explore the dataset in JupyterHub, including:
     - Loading the dataset.
     - Checking the shape of the dataset with `.shape` (to get the number of rows and columns).
     - Using `describe()`, `info()`, and `head()` for basic summaries and previews of the data.

3. **Understanding Methods vs. Attributes**:
   - You asked about the difference between attributes like `df.shape` and methods like `df.describe()`. I explained that attributes represent properties of an object and don't require parentheses, while methods perform an action and need parentheses.

4. **Key Concepts of "Observations" and "Variables"**:
   - You asked about the meanings of "observations" and "variables". I explained that:
     - **Observations** refer to rows (individual data points).
     - **Variables** refer to columns (characteristics or features of the data).

5. **Using `value_counts()`**:
   - We discussed how to use the `value_counts()` function in `pandas` to get the frequency of unique values in categorical columns like `species` or `personality`.

6. **Summary Statistics with `df.describe()`**:
   - Finally, we covered the output of the `describe()` method, which provides summary statistics such as `count`, `mean`, `std`, `min`, `25%`, `50%`, `75%`, and `max`.

Let me know if you'd like to dive deeper into any of these topics or need additional guidance!


## link to interaction: 
https://chatgpt.com/share/6a607a2f-4e4a-424a-8f83-239c88a30149

## Question 7)
1. A "use case" where df.dropna() may be preffered would be where there are more then one column where there is missing data. This may also be used to delete rows of entries where there is null data, 
2. An "opposite use case" where del df['col'] may be preffered might be where there is a specific column that the user wants to delete from the data table, for other reasons then joust null data. 
3. This would prevent entire rows that you may want to keep in the data from being deleted while using the dropna() method. 
4. The removal of all missing data sets depends on the approach you want to take. If the data is concentrated in one column, I would prefer to delete the column while if it is spread out I would use dropna(). 
While if it is a combination of both, I would decide to remove columns with consentrated amounts of null data then delete the individual datasets with null values. 

In [15]:
"""Question 8.1)
df.groupby("col1")["col2"].describe() provides the information on 2 columns rathen then .describe() which only provides info on one. 
"""
titanic.groupby('age')['pclass'].describe()

""" Question 8.2)
While count tallies up the number of present data available, groupby is specific to the respective columns that the user indicates. 
The method "groupby" can provide further knowledge into what data is missing within the column indicated. 
"""

count  mean       std  min   25%  50%   75%  max
age                                                    
0.42     1.0   3.0       NaN  3.0  3.00  3.0  3.00  3.0
0.67     1.0   2.0       NaN  2.0  2.00  2.0  2.00  2.0
0.75     2.0   3.0  0.000000  3.0  3.00  3.0  3.00  3.0
0.83     2.0   2.0  0.000000  2.0  2.00  2.0  2.00  2.0
0.92     1.0   1.0       NaN  1.0  1.00  1.0  1.00  1.0
...      ...   ...       ...  ...   ...  ...   ...  ...
70.00    2.0   1.5  0.707107  1.0  1.25  1.5  1.75  2.0
70.50    1.0   3.0       NaN  3.0  3.00  3.0  3.00  3.0
71.00    2.0   1.0  0.000000  1.0  1.00  1.0  1.00  1.0
74.00    1.0   3.0       NaN  3.0  3.00  3.0  3.00  3.0
80.00    1.0   1.0       NaN  1.0  1.00  1.0  1.00  1.0

[88 rows x 8 columns]

### Chat-log upto this point
Certainly! Here’s a summary of our interaction:

1. **`del df['col']`**: Explained that this command is used to delete a column named `'col'` from a pandas DataFrame `df`.

2. **`df.dropna()`**: Described how this method removes rows or columns from a DataFrame that contain missing values (NaN). It includes details on the parameters like `axis`, `how`, `thresh`, `subset`, and `inplace`.

3. **`df.groupby("col1")["col2"].describe()`**: Clarified that this command groups the DataFrame by the values in column `"col1"`, and then computes descriptive statistics for the column `"col2"` within each group.

Feel free to ask if you need more details or have other questions!

## chat-bot link: 
https://chatgpt.com/share/66e36e97-1d80-8002-890d-1095df268ec0

In [1]:
#Question 8.3 part a) 
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

Comparing whether the chat-bot or google is able to answer this question clearly:
**Chat-bot:** first suggestion is to provide "import pandas as pd" within my code. 
**Google:** The first suggestion is on stack-overflow to provide this text. However, this took some more time to find the answer to as I had to scroll down. 

### Which is easier: Chat-bot. 

In [8]:
#Question 8.3 part b)

titanic.csp

AttributeError: 'DataFrame' object has no attribute 'csp'

Comparing whether the chat-bot or google is able to answer this question clearly:
**Chat-bot:** Suggested to check the file name and if it is spelled correctly. 
**Google:** The first 4 suggested websites did not provide accurate information on how to correct my code. Rather they suggested to change my code to load anyfile inputted and/or unzip files. This did not accurately solve the problem at hand. 

### Which is easier: Chat-bot. 

When changing the typo to titanic.csp, the error became "AttributeError: 'DataFrame' object has no attribute 'csp'"

**Chat-bot:** The chat bot realized I was misspelling "csv" and suggested to check what I spelt.
**Google:** The suggestions I got did not pertain specifically to the case I had. Thus, it was harder to find information on what mistake I was having. However, after researching deeper I came to the conclusion that the mistakes similar people were having were about misspelling the attributes within their code. 

### Which is easier: Chat-bot. 


In [9]:
#Question 8.3 c)
DF.groupby("col1")["col2"].describe() 


NameError: name 'DF' is not defined

**Chat-bot:** suggests that the name provided is not set to any data-set. 
**Google:** Found similar cases where people had the same problem. People indicated that I should make sure to use the same name while working on the same variable 
### Which is easier: Chat-bot. 

In [10]:
#Question 8.3 d)
pd.read_csv(url

SyntaxError: incomplete input (1812998659.py, line 2)

**Chat-bot:** suggested to close bracket, providing the simple solution to this syntax error. 

**Google:** Since this is a vague situations, I found the solution to this problem however it was not specific to python. However, through the search it was clear that the problem with this code was that there must have been a syntax error. 

### Which is easier: Chat-bot. 

In [12]:
#Question 8.3 e)
titanic.groupby("sex")["age"].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

**Chat-bot:** It provided the solution to change the code to "describe" in its third point. 
**Google:** Did not provide the specific solution, but showed that this was a typo error as well.

### Which is easier: Both are equally easy. 

In [14]:
#Question 8.3 f)
titanic.groupby("Sex")["Age"].describe()

KeyError: 'Sex'

**Chat-bot:** Provided mutliple steps to verify the column names and then provided a suggestion to check if names are misspelled. 

**Google:** Was unable to provide any relevent links the first time. I had to shorten the error message repeatedly to find relevent information. 

### Which is easier? Chat-bot. 


In [15]:
#Question 8.3 g)
titanic_df.groupby(sex)["age"].describe()

NameError: name 'titanic_df' is not defined

**Chat-bot:** Providing the error message did not help, since I got an error message stating "NameError". However, while providing the code I recieved an accurate solution to the problem. 

**Google:** I had a similar situation. I only recieved a relevent solution once I inputted my code on the search feature. 

### Which is easier? Neither, both were equally time-consuming. 

**Question 9.**
No I did not. During the lecture I had written down notes on what each section of the material was. 

### Chat-bot log for question 8.3

Certainly! Here's a summary of our interaction:

1. **Issue with CSV Reading**: You encountered an issue with reading a CSV file using `pd.read_csv` and an error related to `'titanics.csv'`. We discussed checking the file name, URL, and method usage to resolve the issue.

2. **Method Error**: You faced an `AttributeError` related to `'csp'`. It was pointed out that `csp` is not a valid method for a `DataFrame`, and the correct method might be `describe`.

3. **Syntax Error**: You encountered a `SyntaxError: incomplete input`, which suggested issues with missing or mismatched syntax elements like parentheses or quotes.

4. **AttributeError with `SeriesGroupBy`**: You had an issue with `describe` not being recognized for a `SeriesGroupBy` object, likely due to a typo in the method name (`describle` should be `describe`).

5. **KeyError**: You faced a `KeyError: 'Sex'`, indicating that the column `'Sex'` does not exist in your DataFrame. We discussed verifying column names and ensuring proper usage of `groupby`.

6. **Final Correction**: You needed to correct the usage of `groupby` with column names as strings and ensure that the column names match those in your DataFrame.

If you have any more questions or need further assistance, feel free to ask!


## Link to ChatBot session:
https://chatgpt.com/share/66e378a5-b158-8002-9389-476af37d701d